lib/make-disk-image.nix

```diff
@@ -13,46 +14,128 @@ let
+    inherit preVM QEMU_OPTS;
+    memSize = 1024;
+  } builder);
+  impure = diskoLib.writeCheckedBash { inherit checked pkgs; } name ''
+    set -efu
+    export PATH=${lib.makeBinPath dependencies}
+    showUsage() {
+    cat <<\USAGE
+    Usage: $script [options]
 
-  memSize = 1024;
-} ''
-  # running udev, stolen from stage-1.sh
-  echo "running udev..."
-  ln -sfn /proc/self/fd /dev/fd
-  ln -sfn /proc/self/fd/0 /dev/stdin
-  ln -sfn /proc/self/fd/1 /dev/stdout
-  ln -sfn /proc/self/fd/2 /dev/stderr
-  mkdir -p /etc/udev
-  ln -sfn ${systemToInstall.config.system.build.etc}/etc/udev/rules.d /etc/udev/rules.d
-  mkdir -p /dev/.mdadm
-  ${pkgs.systemdMinimal}/lib/systemd/systemd-udevd --daemon
-  udevadm trigger --action=add
-  udevadm settle
+    Options:
+    * --preDiskoFiles <src> <dst>
+      copies the src to the dst on the VM, before disko is run
+      This is useful to provide secrets like LUKS keys, or other files you need for formating
+    * --postDiskoFiles <src> <dst>
+      copies the src to the dst on the finished image.
+      These end up in the images later and is useful if you want to add some extra stateful files
+    USAGE
+    }
 
-  # populate nix db, so nixos-install doesn't complain
-  export NIX_STATE_DIR=$TMPDIR/state
-  nix-store --load-db < ${pkgs.closureInfo {
-    rootPaths = [ systemToInstall.config.system.build.toplevel ];
-  }}/registration
+    export out=$PWD
+    TMPDIR=$(mktemp -d); export TMPDIR
+    trap 'rm -rf "$TMPDIR"' EXIT
+    cd "$TMPDIR"
 
-  ${systemToInstall.config.system.build.diskoScript}
-  ${pkgs.nixos-install-tools}/bin/nixos-install --system ${systemToInstall.config.system.build.toplevel} --keep-going --no-channel-copy -v --no-root-password --option binary-caches ""
-'')
+    touch copy_to_xchg
+    mkdir copy_before_disko
+    mkdir copy_after_disko
+
+    while [[ $# -gt 0 ]]; do
+      case "$1" in
+      --pre-format-files)
+        src=$2
+        dst=$3
+        cp -r "$src" copy_before_disko/"$(echo "$dst" | base64)"
+        shift 2
+        ;;
+      --post-format-files)
+        src=$2
+        dst=$3
+        cp -r "$src" copy_after_disko/"$(echo "$dst" | base64)"
+        shift 2
+        ;;
+      *)
+        showUsage
+        exit 1
+        ;;
+      esac
+      shift
+    done
+
+    export preVM=${diskoLib.writeCheckedBash { inherit pkgs checked; } "preVM.sh" ''
+      set -efu
+      mv copy_before_disko xchg/
+      mv copy_after_disko xchg/
+      ${preVM}
+    ''}
+    export postVM=${diskoLib.writeCheckedBash { inherit pkgs checked; } "postVM.sh" postVM}
+    export origBuilder=${pkgs.writeScript "disko-builder" ''
+      set -eu
+      export PATH=${lib.makeBinPath dependencies}
+      for src in /tmp/xchg/copy_before_disko/*; do
+        dst=$(basename "$src" | base64 -d)
+        mkdir -p "$(dirname "$dst")"
+        cp -r "$src" "$dst"
+      done
+      set -f
+      ${builder}
+      set +f
+      for dir in /tmp/xchg/copy_after_disko/*; do
```

Comment:
If this /tmp directory is outside of a nix build than it should rather use mktmp. Otherwise it will collide if multiple concurrent builds happen.

Reply:
ok. looks like this is in the vm.